// ungit/components/remotes/remotes.bundle.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){

var ko = require('knockout');
var _ = require('lodash');
var async = require('async');
var components = require('ungit-components');
var programEvents = require('ungit-program-events');

components.register('remotes', function(args) {
  return new RemotesViewModel(args.server, args.repoPath);
});

function RemotesViewModel(server, repoPath) {
  var self = this;
  this.repoPath = repoPath;
  this.server = server;
  this.remotes = ko.observable([]);
  this.currentRemote = ko.observable(null);
  this.currentRemote.subscribe(function(value) {
    programEvents.dispatch({ event: 'current-remote-changed', newRemote: value });
  });
  this.fetchLabel = ko.computed(function() {
    if (self.currentRemote()) return 'Fetch from ' + self.currentRemote();
    else return 'No remotes specified';
  })

  this.fetchingProgressBar = components.create('progressBar', { predictionMemoryKey: 'fetching-' + this.repoPath(), temporary: true });

  this.fetchEnabled = ko.computed(function() {
    return self.remotes().length > 0;
  });

  this.shouldAutoFetch = ungit.config.autoFetch;
  this.updateRemotes();
}
RemotesViewModel.prototype.updateNode = function(parentElement) {
  ko.renderTemplate('remotes', this, {}, parentElement);
}
RemotesViewModel.prototype.clickFetch = function() { this.fetch({ nodes: true, tags: true }); }
RemotesViewModel.prototype.onProgramEvent = function(event) {
  if (event.event == 'request-credentials') this.fetchingProgressBar.pause();
  else if (event.event == 'request-credentials-response') this.fetchingProgressBar.unpause();
  else if (event.event == 'request-fetch-tags') this.fetch({ tags: true });
}
RemotesViewModel.prototype.fetch = function(options) {
  if (this.fetchingProgressBar.running()) return;
  var self = this;

  this.fetchingProgressBar.start();
  var jobs = [];
  if (options.tags) jobs.push(function(done) { self.server.get('/remote/tags', { path: self.repoPath(), remote: self.currentRemote() }, done); });
  if (options.nodes) jobs.push(function(done) { self.server.post('/fetch', { path: self.repoPath(), remote: self.currentRemote() }, done);  });
  async.parallel(jobs, function(err, result) {
    self.fetchingProgressBar.stop();

    if (!err && options.tags) programEvents.dispatch({ event: 'remote-tags-update', tags: result[0] });
  });
}

RemotesViewModel.prototype.updateRemotes = function() {
  var self = this;
  this.server.get('/remotes', { path: this.repoPath() }, function(err, remotes) {
    if (err && err.errorCode == 'not-a-repository') return true;
    if (err) return;
    remotes = remotes.map(function(remote) {
      return {
        name: remote,
        changeRemote: function() { self.currentRemote(remote) }
      }
    });
    self.remotes(remotes);
    if (!self.currentRemote() && remotes.length > 0) {
      if (_.find(remotes, { 'name': 'origin' })) // default to origin if it exists
        self.currentRemote('origin');
      else // otherwise take the first one
        self.currentRemote(remotes[0].name);
      if (self.shouldAutoFetch) {
        self.fetch({ nodes: true, tags: true });
      }
    }
    self.shouldAutoFetch = false;
  });
}
RemotesViewModel.prototype.showAddRemoteDialog = function() {
  var self = this;
  var diag = components.create('addremotedialog');
  diag.closed.add(function() {
    if (diag.isSubmitted()) {
      self.server.post('/remotes/' + encodeURIComponent(diag.name()), { path: self.repoPath(), url: diag.url() }, function(err, res) {
        if (err) return;
        self.updateRemotes();
      })
    }
  });
  programEvents.dispatch({ event: 'request-show-dialog', dialog: diag });
}

RemotesViewModel.prototype.remoteRemove = function(remote) {
  var self = this;
  var diag = components.create('yesnodialog', { title: 'Are you sure?', details: 'Deleting ' + remote.name + ' remote cannot be undone with ungit.'});
  diag.closed.add(function() {
    if (diag.result()) {
      self.fetchingProgressBar.start();
      self.server.del('/remotes/' + remote.name, { path: self.repoPath() }, function(err, result) {
        if (err) {
          console.log(err);
          return;
        }

        self.updateRemotes();
        self.fetchingProgressBar.stop();
      });
    }
  });
  programEvents.dispatch({ event: 'request-show-dialog', dialog: diag });
}

},{"async":"async","knockout":"knockout","lodash":"lodash","ungit-components":"ungit-components","ungit-program-events":"ungit-program-events"}]},{},[1])
