// ungit/src/bugtracker.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var winston = require('winston');
var sysinfo = require('./sysinfo');
var config = require('./config');
var raven = require('raven');
var client = new raven.Client('https://58f16d6f010d4c77900bb1de9c02185f:84b7432f56674fbc8522bc84cc7b30f4@app.getsentry.com/12434');

var BugTracker = function () {
  function BugTracker(subsystem) {
    var _this = this;

    _classCallCheck(this, BugTracker);

    if (!config.bugtracking) return;

    this.subsystem = subsystem;
    this.appVersion = 'unknown';
    this.userHash = 'unkown';

    sysinfo.getUngitVersion(function (err, ungitVersion) {
      _this.appVersion = ungitVersion;
      winston.info('BugTracker set version: ' + _this.appVersion);
    });

    sysinfo.getUserHash(function (err, userHash) {
      _this.userHash = userHash;
      winston.info('BugTracker set user hash');
    });
  }

  _createClass(BugTracker, [{
    key: 'notify',
    value: function notify(exception, clientName, callback) {
      if (!config.bugtracking) return;

      var options = {
        user: { id: this.userHash },
        tags: {
          version: this.appVersion,
          subsystem: this.subsystem,
          deployment: config.desktopMode ? 'desktop' : 'web'
        }
      };

      client.captureException(exception, options);
    }
  }]);

  return BugTracker;
}();

module.exports = BugTracker;
