// ungit/src/ungit-plugin.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var fs = require('fs');
var path = require('path');
var async = require('async');
var express = require('express');
var winston = require('winston');
var config = require('./config');

var assureArray = function assureArray(obj) {
  return Array.isArray(obj) ? obj : [obj];
};

var UngitPlugin = function () {
  function UngitPlugin(args) {
    _classCallCheck(this, UngitPlugin);

    this.dir = args.dir;
    this.path = args.path;
    this.httpBasePath = args.httpBasePath;
    this.manifest = JSON.parse(fs.readFileSync(path.join(this.path, "ungit-plugin.json")));
    this.name = this.manifest.name || this.dir;
    this.config = config.pluginConfigs[this.name] || {};
  }

  _createClass(UngitPlugin, [{
    key: 'init',
    value: function init(env) {
      if (this.manifest.server) {
        var serverScript = require(path.join(this.path, this.manifest.server));
        serverScript.install({
          app: env.app,
          httpServer: env.httpServer,
          ensureAuthenticated: env.ensureAuthenticated,
          ensurePathExists: env.ensurePathExists,
          git: require('./git-promise'),
          config: env.config,
          socketIO: env.socketIO,
          socketsById: env.socketsById,
          pluginConfig: this.config,
          httpPath: env.pathPrefix + '/plugins/' + this.name,
          pluginApiVersion: require('../package.json').ungitPluginApiVersion
        });
      }
      env.app.use('/plugins/' + this.name, express.static(this.path));
    }
  }, {
    key: 'compile',
    value: function compile(callback) {
      var _this = this;

      winston.info('Compiling plugin ' + this.path);
      var exports = this.manifest.exports || {};
      var tasks = [];

      if (exports.raw) {
        var raw = assureArray(exports.raw);
        raw.forEach(function (rawSource) {
          tasks.push(function (callback) {
            fs.readFile(path.join(_this.path, rawSource), function (err, text) {
              callback(err, text + '\n');
            });
          });
        });
      }

      if (exports.javascript) {
        var js = assureArray(exports.javascript);

        js.forEach(function (filename) {
          tasks.push(function (callback) {
            callback(null, '<script type="text/javascript" src="' + config.rootPath + '/plugins/' + _this.name + '/' + filename + '"></script>');
          });
        });
      }

      if (exports.knockoutTemplates) {
        Object.keys(exports.knockoutTemplates).forEach(function (templateName) {
          tasks.push(function (callback) {
            fs.readFile(path.join(_this.path, exports.knockoutTemplates[templateName]), function (err, text) {
              callback(err, '<script type="text/html" id="' + templateName + '">\n' + text + '\n</script>\n');
            });
          });
        });
      }

      if (exports.css) {
        var css = assureArray(exports.css);
        css.forEach(function (cssSource) {
          tasks.push(function (callback) {
            callback(null, '<link rel="stylesheet" type="text/css" href="' + config.rootPath + '/plugins/' + _this.name + '/' + cssSource + '" />');
          });
        });
      }

      async.parallel(tasks, function (err, result) {
        if (err) throw err;
        callback(err, '<!-- Component: ' + _this.name + ' -->\n' + result.join('\n'));
      });
    }
  }]);

  return UngitPlugin;
}();

module.exports = UngitPlugin;
